### ISSUE TITLE
Promises run only after closing the window

### ISSUE THREAD

**Issue by wstam88** [romgrk/node-gtk#121]
For some reason promises do not resolve until the window is closed.

For example when adding this piece of code just before `win.showAll();` at examples/builder.js.

```js
let myPromise = function(){
    return new Promise((resolve, reject) => {
        resolve('Hi...');
        console.log('This gets printed just fine...')
    });
}

myPromise().then(res => {
    console.log('this only runs after closing/exit the window');
    label.setText(res);
});
```

Doesn't really matter where I place it, also happens inside the body of an event callback function for example.

**Comment by romgrk** [comment]
Ok, this is weird.

So normally we're already handling micro-tasks [here](https://github.com/romgrk/node-gtk/blob/master/src/util.cc#L49-L61), so promises should be called even when GLib's loop is running (`Gtk.main()`).

**Comment by wstam88** [comment]
Well if you put it in an interval it does work! And the other ones outside the interval suddenly also work 🤔

```js
setInterval(() => {
    myPromise().then(res => {
        console.log('works...')
    });
}, 1000)
```

**Comment by wstam88** [comment]
Sorry didn't understood your comment correctly...

**Comment by subjectdenied** [comment]
i'm not sure we mean the same, but inside event-bodies i suceeded by using nodejs setImmediate

**Comment by wstam88** [comment]
I noticed that as well, but it doesn't really run immediately or consistently. Sometimes it takes a few seconds and sometimes it just skips. 

As soon as [originalMain()](https://github.com/romgrk/node-gtk/blob/c82cd2639df86855f59e4fc5cad63af350f53613/lib/overrides/Gtk-3.0.js#L25) is called it completely blocks. 

The promise status is `resolved` and the value is there, it's just the callback that gets stuck in the eventloop/next_tick. When [originalMain()](https://github.com/romgrk/node-gtk/blob/c82cd2639df86855f59e4fc5cad63af350f53613/lib/overrides/Gtk-3.0.js#L25) is stopped then suddenly every promise resolve callback is called.

Also setTimeout inside an event callback doesn't run until closed.
```js
btn2.on('clicked', () => {
    // this runs, and promise is resolved
    let promy = prom();

    let timeout = setTimeout(() => {
        console.log('this does not run...')
    }, 1000);

    debugger

    // it's just the callback that gets stuck
    promy.then((val) => {
        // this block runs after closing the Gtk window
        console.log(val);
    });
});
```

![example](https://i.imgur.com/Esi3aeU.gif)

**Comment by romgrk** [comment]
I've made things work as expected with #121 (note my skillful avoidance of the word "fixed"), I'm waiting for the tests to complete before merging.

**Comment by wstam88** [comment]
See my comment [here](https://github.com/romgrk/node-gtk/pull/122#issuecomment-500100950)

**Comment by romgrk** [comment]
The point of the interval was not to trigger an event-loop iteration each time it is called, but rather to keep it running by having something on it, which doesn't seem to be working. I'll try to have a minimal reproduction of the issue and will investigate in more details what's going on in libuv & GLib loops.

**Comment by romgrk** [comment]
I also note that for axios, if one clicks the "Action" button in the new `builder.js` version, the axios promise is processed. It seems nodejs doesn't notice that some microtaks have been queued.

**Comment by haijin-development** [comment]
Hi.

I had this issue too.
There is a workaround that seems to fix it.
If an interval runs additional iterations of the Gtk main loop it seems to both resolve the pending promises and process the Gtk events queue.

So instead of calling

```javascript
Gtk.main()
```
I am doing

```javascript
let interval
interval = setInterval( () => {
  if( runMainLoop !== true ) { clearInterval(interval) }
  Gtk.mainIterationDo(false)
}, 0)
Gtk.main()
```

Would it be possible to simplify the main loop without blocking node events loop with

```javascript
let runMainLoop = true
Gtk.main = function() {
   let interval
   interval = setInterval( () => {
     if( runMainLoop !== true ) { clearInterval(interval) }
     Gtk.mainIterationDo(false)
   }, 0)
}
```
? (it is a question I am asking, not the aswer)

**Comment by romgrk** [comment]
Hey,

Thanks for your input! It is certainly a nice workaround to investigate, I am however currently too short on time to take a look. If you're interested in investigating and opening a PR, it would be very welcome. Otherwise, I'll take a look eventually but I don't know when :/

Thanks

**Comment by haijin-development** [comment]
Hi.
Thanks for answering.
I will try to take a look at it but I haven't done any low level development in quite some time and never did bindings to node so I would need to take some time to catch up before requesting a PR.

Thanks and happy hollidays.

**Comment by Drulac** [comment]
Hello :)
Same issue here :sweat_smile: 

@haijin-development I'm not able to use your workaround :sweat_smile: 

```
window.on('show', async () => {
                let interval
                runMainLoop = true
                interval = setInterval(() => {
                        if (runMainLoop !== true) {
                                clearInterval(interval)
                        }
                        Gtk.mainIterationDo(false)
                }, 1)
                Gtk.main()
        })
```
It should be working ?

**Comment by Drulac** [comment]
Ok, I've installed node 10.18.0 (I was previously using node 11, I'v tested node 12 not working, and node 13 I can't compile `node-gtk`)  and works :woman_shrugging:

**Comment by Drulac** [comment]
I've found a strange workaround : running a `setInterval` before `Gtk.main()`
He will be correctly fired, and if I start another interval (or promise) from the inside of this interval, it's working

**Comment by romgrk** [comment]
Thanks for the feedback, if you can open a separate issue for the crashes (with the steps to reproduce) it would be nice. I should be getting back to this issue soon, I had a few other things to fix first.

**Comment by romgrk** [comment]
Got the issue fixed on Linux, but I still need to debug it on macOS before merging the PR.

**Comment by strivelabs-pablorb** [comment]
Hello, sorry for disturbing you. What is the state of this? Has this been fixed? Because currently on a macOS M1 using the first sync example from @Drulac does not work for me. In fact, it freezes and I have to force quit the app.
```
Gtk.main()
console.log('yeah')
```
If I only use Gtk.init() the windows appear but the console.log gets fired after I close the window.

**Comment by romgrk** [comment]
That's the expected behavior. `gtk_main()` does not return until the GLib event loop has terminated, which means when the GTK window closes.

If you want to run code after the window is opened, you can look at the examples for Gtk4. I think the `activated` signal would be what you're looking for, but confirm with the gtk docs.

**Comment by strivelabs-pablorb** [comment]
Ill check that code. Thx for the answer. To give more context we believe this issue is what is preventing our app from closing when the pc restarts.
